Rule Change: support `Array.fromAsync` in `no-invalid-this`

### What rule do you want to change?

no-invalid-this

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint no-invalid-this: "error" */

const contents = await Array.fromAsync(
    ["file1.txt", "file2.txt"],
    async function (file) {
        return await this.read(file);
    },
    { read: file => readFile(file, 'utf8') }
);
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWludmFsaWQtdGhpczogXCJlcnJvclwiICovXG5cbmNvbnN0IGNvbnRlbnRzID0gYXdhaXQgQXJyYXkuZnJvbUFzeW5jKFxuICAgIFtcImZpbGUxLnR4dFwiLCBcImZpbGUyLnR4dFwiXSxcbiAgICBhc3luYyBmdW5jdGlvbiAoZmlsZSkge1xuICAgICAgICByZXR1cm4gYXdhaXQgdGhpcy5yZWFkKGZpbGUpO1xuICAgIH0sXG4gICAgeyByZWFkOiBmaWxlID0+IHJlYWRGaWxlKGZpbGUsICd1dGY4JykgfVxuKTtcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==
)

### What does the rule currently do for this code?

A problem is reported:

```text
  6:22  error  Unexpected 'this'  no-invalid-this
```

### What will the rule do after it's changed?

No problem will be reported.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

[`Array.fromAsync`](https://github.com/tc39/proposal-array-from-async#readme) has reached TC39 stage 4 in ES2026. It follows the same signature and similar semantics as `Array.from`, so I think the rule should treat it in the same way when handling `this` in the callback function. See also #20438.

--- a/lib/rules/utils/ast-utils.js
+++ b/lib/rules/utils/ast-utils.js
@@ -527,6 +527,15 @@ function isArrayFromMethod(node) {
 }
 
 /**
+ * Checks whether or not a node is `Array.fromAsync`.
+ * @param {ASTNode} node A node to check.
+ * @returns {boolean} Whether or not the node is a `Array.fromAsync`.
+ */
+function isArrayFromAsyncMethod(node) {
+	return isSpecificMemberAccess(node, "Array", "fromAsync");
+}
+
+/**
  * Checks whether or not a node is a method which expects a function as a first argument, and `thisArg` as a second argument.
  * @param {ASTNode} node A node to check.
  * @returns {boolean} Whether or not the node is a method which expects a function as a first argument, and `thisArg` as a second argument.
@@ -1462,6 +1471,7 @@ module.exports = {
 	isLoop,
 	isInLoop,
 	isArrayFromMethod,
+	isArrayFromAsyncMethod,
 	isParenthesised,
 	createGlobalLinebreakMatcher,
 	equalTokens,
@@ -1793,7 +1803,10 @@ module.exports = {
 							isNullOrUndefined(parent.arguments[1])
 						);
 					}
-					if (isArrayFromMethod(parent.callee)) {
+					if (
+						isArrayFromMethod(parent.callee) ||
+						isArrayFromAsyncMethod(parent.callee)
+					) {
 						return (
 							parent.arguments.length !== 3 ||
 							parent.arguments[1] !== currentNode ||

--- a/tests/lib/rules/no-eval.js
+++ b/tests/lib/rules/no-eval.js
@@ -135,6 +135,10 @@ ruleTester.run("no-eval", rule, {
 		"array.findLast(function (x) { return this.eval.includes(x); }, { eval: ['foo', 'bar'] });",
 		"callbacks.findLastIndex(function (cb) { return cb(this.eval); }, this);",
 		"['1+1'].flatMap(function (str) { return this.eval(str); }, new Evaluator);",
+		{
+			code: "Array.fromAsync(values, async function (value) { return this.eval(await value); }, context);",
+			languageOptions: { ecmaVersion: 2017 },
+		},
 
 		// Allows indirect eval
 		{ code: "(0, eval)('foo')", options: [{ allowIndirect: true }] },

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -481,13 +481,13 @@ const patterns = [
 	},
 
 	// Array methods.
-	{
-		code: "Array.from([], function() { console.log(this); z(x => console.log(x, this)); });",
+	...["from", "fromAsync"].map(methodName => ({
+		code: `Array.${methodName}([], function() { console.log(this); z(x => console.log(x, this)); });`,
 		languageOptions: { ecmaVersion: 6 },
 		errors,
 		valid: [NORMAL],
 		invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
-	},
+	})),
 	...[
 		"every",
 		"filter",
@@ -506,12 +506,12 @@ const patterns = [
 		valid: [NORMAL],
 		invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
 	})),
-	{
-		code: "Array.from([], function() { console.log(this); z(x => console.log(x, this)); }, obj);",
+	...["from", "fromAsync"].map(methodName => ({
+		code: `Array.${methodName}([], function() { console.log(this); z(x => console.log(x, this)); }, obj);`,
 		languageOptions: { ecmaVersion: 6 },
 		valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
 		invalid: [],
-	},
+	})),
 	...[
 		"every",
 		"filter",
@@ -1124,6 +1124,17 @@ ruleTesterTypeScript.run("no-invalid-this", rule, {
         `,
 
 		`
+    Array.fromAsync(
+      [],
+      function () {
+        console.log(this);
+        z(x => console.log(x, this));
+      },
+      obj,
+    );
+        `,
+
+		`
     foo.every(function () {
       console.log(this);
       z(x => console.log(x, this));
@@ -1598,6 +1609,15 @@ ruleTesterTypeScript.run("no-invalid-this", rule, {
 		{
 			code: `
     Array.from([], function () {
+      console.log(this);
+      z(x => console.log(x, this));
+    });
+          `,
+			errors,
+		},
+		{
+			code: `
+    Array.fromAsync([], function () {
       console.log(this);
       z(x => console.log(x, this));
     });

